### PR TITLE
Add get_balance api for RPC

### PR DESCRIPTION
**Pull request by mikedoan**
Add get_balance api for RPC

### CODE REVIEW

File: hmy/api_backend.go [hunk 1]
```diff
@@ -185,3 +198,9 @@ func (b *APIBackend) GetPoolTransactions() (types.Transactions, error) {
 	}
 	return txs, nil
 }
+
+// GetBalance returns balance of an given address.
+func (b *APIBackend) GetBalance(address common.Address) (*hexutil.Big, error) {
+	balance, err := b.hmy.nodeAPI.GetBalanceOfAddress(address)
+	return (*hexutil.Big)(balance), err
```
Comment (LeoHChen) [inline hmy/api_backend.go]: is there any risk (balance) be nil?

Comment (mikedoan) [reply hmy/api_backend.go]: Get balance is handled by ethereum GetBalance. I look around ethereum code if nil it returns Big0 etc... For extremely conservative, we can have that check. But if we do that we would have to handle that many other places.
```
// GetBalance retrieves the balance from the given address or 0 if object not found
func (stateDB *DB) GetBalance(addr common.Address) *big.Int {
	stateObject := stateDB.getStateObject(addr)
	if stateObject != nil {
		return stateObject.Balance()
	}
	return common.Big0
}
```

Comment (LeoHChen) [reply hmy/api_backend.go]: fine

File: hmy/api_backend.go [hunk 2]
```diff
@@ -127,53 +130,63 @@ func (b *APIBackend) ProtocolVersion() int {
...
-// SubscribeChainEvent ...
+// SubscribeChainEvent subcribes chain event.
+// TODO: this is not implemented or verified yet for harmony.
 func (b *APIBackend) SubscribeChainEvent(ch chan<- core.ChainEvent) event.Subscription {
 	return b.hmy.BlockChain().SubscribeChainEvent(ch)
 }
 
-// SubscribeChainHeadEvent ...
+// SubscribeChainHeadEvent subcribes chain head event.
+// TODO: this is not implemented or verified yet for harmony.
```
Comment (lzl124631x) [inline hmy/api_backend.go]: I think some of these are already working now, as I can listen to filterChange of newBlockFilter

File: hmy/api_backend.go [hunk 3]
```diff
@@ -127,53 +130,63 @@ func (b *APIBackend) ProtocolVersion() int {
...
 
 // HeaderByHash ...
+// TODO: this is not implemented or verified yet for harmony.
 func (b *APIBackend) HeaderByHash(ctx context.Context, blockHash common.Hash) (*types.Header, error) {
 	// TODO(ricl): implement
 	return nil, nil
 }
 
 // ServiceFilter ...
+// TODO: this is not implemented or verified yet for harmony.
```
Comment (lzl124631x) [inline hmy/api_backend.go]: For the APIs that I've already added TODO, no need to add the TODO twice.

File: internal/hmyapi/blockchain.go [hunk 4]
```diff
@@ -72,9 +72,5 @@ func (s *PublicBlockChainAPI) GetStorageAt(ctx context.Context, address common.A
 // given block number. The rpc.LatestBlockNumber and rpc.PendingBlockNumber meta
 // block numbers are also allowed.
 func (s *PublicBlockChainAPI) GetBalance(ctx context.Context, address common.Address, blockNr rpc.BlockNumber) (*hexutil.Big, error) {
-	state, _, err := s.b.StateAndHeaderByNumber(ctx, blockNr)
-	if state == nil || err != nil {
-		return nil, err
-	}
-	return (*hexutil.Big)(state.GetBalance(address)), state.Error()
+	return s.b.GetBalance(address)
```
Comment (lzl124631x) [inline internal/hmyapi/blockchain.go]: Why you just ignored the blockNr? That's an important input parameter.

Comment (lzl124631x) [reply internal/hmyapi/blockchain.go]: Have you tested the API of the old code and new code? What are the behaviors?

Comment (mikedoan) [reply internal/hmyapi/blockchain.go]: agree. right now for testing we will try to get the latest balance first. We will add complete behaviour later. Let me add todo

File: hmy/api_backend.go (hunk 3 above)
Comment (mikedoan) [reply hmy/api_backend.go]: done

**Comment by lzl124631x** [comment]
I tested locally that the getBalance is working but this error I guess is because we don't have that account in blockchain.

image

File: internal/hmyapi/blockchain.go (hunk 4 above)
Comment (lzl124631x) [reply internal/hmyapi/blockchain.go]: I suspect the old code works.

**Comment by mikedoan** [comment]
Richard please approve.

**Comment by mikedoan** [comment]
Screen Shot 2019-05-30 at 11 49 26 AM

**Pull request merged**